Feature: Dominant Eigenvalue Estimator (#750)
LLNL/sundials: doc/shared/sundomeigest/SUNDomEigEst_Arnoldi.rst

approximate some of the eigenvalues of :math:`A`; the dominant eigenvalue of :math:`A` is
well-approximated by the dominant eigenvalue of :math:`H_m`.

Arnoldi iteration works for matrices with both real and complex eigenvalues.  It supports
estimations with a user-specified fixed Krylov subspace dimension (at least 3).  While
the choice of dimension results in a prefixed amount of memory, it strictly
determines how good an estimation is.  To improve the estimation accuracy, we have found that 
preprocessing set by :c:func:`SUNDomEigEst_SetNumPreProcess` is particularly useful.  
This operation is free from any additional memory requirement and is further explained below.

The matrix :math:`A` is not required explicitly; only a routine that provides an 
approximation of the matrix-vector product, :math:`Av`, is required.

.. _SUNDomEigEst.ARNOLDI.Usage:

SUNDomEigEst_ARNOLDI Usage
-------------------------------

The header file to be included when using this module is ``sundomeigest/sundomeigest_arnoldi.h``.
The SUNDomEigEst_ARNOLDI module is accessible from all SUNDIALS solvers *without* linking to the

> ``libsundials_sundomeigestarnoldi`` module library.

Reviewer: Copying from previous PR:
@maggul asked "@gardner48, is this still a true statement after your adjustments in cmakes?"

The header file to be included when using this module is ``sundomeigest/sundomeigest_arnoldi.h``.
The SUNDomEigEst_ARNOLDI module is accessible from all SUNDIALS solvers *without* linking to the

> ``libsundials_sundomeigestarnoldi`` module library after enabling SUNDIALS interfaces to the LAPACK library.

Reviewer: Whether the previous paragraph is still accurate or not, the text is duplicated here, so something needs to be removed.

> This LAPACK dependence is limited to the eigenvalue estimation of the Hessenberg matrix using the 
> ``dgeev``` and/or ``sgeev`` functions.

Reviewer: I recommend that we remove this sentence, since it's likely that `dgeev/sgeev` themselves call other LAPACK/BLAS functions.  I recommend that users just link to a full LAPACK library, instead of somehow linking to only a few pieces.
```suggestion
```

The module SUNDomEigEst_ARNOLDI provides the following user-callable routines:

.. c:function:: SUNDomEigEstimator SUNDomEigEst_Arnoldi(N_Vector q, int kry_dim, int num_warmups, SUNContext sunctx);

   This constructor function creates and allocates memory for an Arnoldi
   ``SUNDomEigEstimator``.

   **Arguments:**

>       * *q* -- a template vector.

Reviewer: Is `q` really just a template vector, or it it actually very first initial guess at the dominant eigenvector?  If it's the initial guess, then that should be explained here, as well as a warning that users should ensure that this is not exactly a non-dominant eigenvector of the Jacobian (e.g., they can use a random vector).

Author: Please see the correction.

      * *kry_dim* -- the dimension of the Krylov subspaces.
      * *num_warmups* -- number of preprocessing warmups.
      * *sunctx* -- the :c:type:`SUNContext` object (see :numref:`SUNDIALS.SUNContext`)

   **Return value:**
      If successful, a ``SUNDomEigEstimator`` object.  If *q* is
      incompatible, this routine will return ``NULL``.

   **Notes:**
      This routine will perform consistency checks to ensure that it is
      called with a consistent ``N_Vector`` implementation (i.e.  that it
      supplies the requisite vector operations).

      A ``kry_dim`` argument that is :math:`\leq 2` will result in the default
      value (3).  This default value is particularly chosen to minimize the memory
      footprint.

      A ``num_warmups`` argument that is :math:` < 0` will result in the default
      value (100).  This default value is particularly chosen to minimize the memory

>       footprint by lowering the required ``kry_dim``.

Reviewer: Please add a comment regarding whether these warmups are performed before every "Estimate" call, or only once.

Author: Whenever we call estimate, this warmups are performed. If we consider DEE as stand alone module, this makes sense since we don't know if the last ``q`` vector will be a good guess for the next estimation, if the user ever need to call the estimate one more time. 

On the other hand, if the user is a SUNDIALS integrator, it has an integer ``num_succ_warmups`` in the step memory, which would be modified by a set routine within the integrator. Once we initialize the DEE and use the ``num_warmups`` for the very first preprocess, integrator calls the `SUNDomEigEst_SetNumPreProcess` with ``num_succ_warmups`` to ensure succeeding warmups are performed ``num_succ_warmups`` times. Default for ``num_succ_warmups`` is 0. That makes sense, since we expect the last ``q`` vector to be a good guess. But for some reason, maybe to allow less frequent DEE estimations, or expecting a significant change from one estimation to the other, user is allowed to take as many preprocess as they wish by changing the default value. To that end they need to call `LSRKStepSetNumSucceedingWarmups`. This are all stored in the integrator level and explained in the corresponding part. 

To wrap up, the statement that  warmups are performed before every "Estimate" is true for DEE as a stand alone module. On the other hand, it is not true if it's attached to an integrator since we do preprocess once there and let the user decide what should happen in the next estimate through integrator set routines (or no warmups in default). I am not sure in what extend I should comment here. I would appreciate any suggestion.

determines how good an estimation is.  To improve the estimation accuracy, we have found that 
preprocessing set by :c:func:`SUNDomEigEst_SetNumPreProcess` is particularly useful.  
This operation is free from any additional memory requirement and is further explained below.

The matrix :math:`A` is not required explicitly; only a routine that provides an 
approximation of the matrix-vector product, :math:`Av`, is required.

.. _SUNDomEigEst.ARNOLDI.Usage:

SUNDomEigEst_ARNOLDI Usage
-------------------------------

The header file to be included when using this module is ``sundomeigest/sundomeigest_arnoldi.h``.
The SUNDomEigEst_ARNOLDI module is accessible from all SUNDIALS solvers *without* linking to the
``libsundials_sundomeigestarnoldi`` module library after enabling SUNDIALS interfaces to the LAPACK 
library.

The module SUNDomEigEst_ARNOLDI provides the following user-callable routines:

.. c:function:: SUNDomEigEstimator SUNDomEigEst_Arnoldi(N_Vector q, int kry_dim, int num_warmups, SUNContext sunctx);

   This constructor function creates and allocates memory for an Arnoldi
   ``SUNDomEigEstimator``.

   **Arguments:**

>       * *q* -- the initial guess for the dominant eigenvector -- It must avoid being a non-dominant eigenvector of the Jacobian.

Reviewer: ```suggestion
      * *q* -- the initial guess for the dominant eigenvector; this should not be a non-dominant eigenvector of the Jacobian.
```

      * *kry_dim* -- the dimension of the Krylov subspaces.
      * *num_warmups* -- number of preprocessing warmups.
      * *sunctx* -- the :c:type:`SUNContext` object (see :numref:`SUNDIALS.SUNContext`)

   **Return value:**
      If successful, a ``SUNDomEigEstimator`` object.  If *q* is
      incompatible, this routine will return ``NULL``.

   **Notes:**
      This routine will perform consistency checks to ensure that it is
      called with a consistent ``N_Vector`` implementation (i.e.  that it
      supplies the requisite vector operations).

      A ``kry_dim`` argument that is :math:`\leq 2` will result in the default
      value (3).  This default value is particularly chosen to minimize the memory
      footprint.

      A ``num_warmups`` argument that is :math:` < 0` will result in the default
      value (100).  This default value is particularly chosen to minimize the memory
      footprint by lowering the required ``kry_dim``.

>       Whenever :c:func:`SUNDomEig_Estimate` is called, ``num_warmups`` times warmups are 
>       performed. This makes sense when considering the DEE module standalone since it cannot assume 
>       that the last ``q`` vector will be a good initial guess for the next estimation. Therefore, 
>       if the user needs to call :c:func:`SUNDomEig_Estimate` again, warmups help ensure accuracy. 
>       However, when DEE is used within an integrator, there is a ``num_succ_warmups`` field in its 
>       step memory. This field is updated by a set function within the integrator, e.g., by 
>       :c:func:`LSRKStepSetNumSucceedingWarmups`. Once DEE is initialized and ``num_warmups`` is used 
>       for the first preprocessing, the integrator calls :c:func:`SUNDomEigEst_SetNumPreProcess` 
>       internally to ensure that all succeeding warmups are performed ``num_succ_warmups`` times.

Reviewer: ```suggestion
      When :c:func:`SUNDomEig_Estimate` is called, then prior to beginning the Arnoldi
      process, ``num_warmups`` power iterations are performed on ``q`` to generate an
      improved initial guess.  However, when the DEE is used in a time-dependent 
      context, it is likely that the most-recent ``q`` will provide a suitable initial 
      guess for the subsequent call to :c:func:`SUNDomEig_Estimate`.  Thus, when the DEE 
      is used by LSRKStep (see :c:func:`LSRKStepSetDomEigEstimator`), the initial 
      value of ``num_warmups`` will be overwritten after the first 
      :c:func:`SUNDomEig_Estimate` call (see :c:func:`LSRKStepSetNumSucceedingWarmups`).
```
